docs/adr/0001-backup-strategy.md

```diff
@@ -0,0 +1,56 @@
+# Backup strategy
+
+Date: 2024-12
+
+## Status
+
+Draft
+
+I'm proposing that we keep this in draft until backup-related issues in the
+Onboarding OpenCodelists initiative are resolved or deferred. Then the outcomes
+can be combined into one ADR, which will be easier to follow in future.
+
+## Context
+
+OpenCodelists runs on a DigitalOcean droplet with Dokku-managed docker
+containers serving Django apps. There are multiple SQLite databases. The
+storage file system is local. Digital Ocean offer Block and Object remote
+storage solutions but we don't currently use those.
+
+The droplet is backed up weekly, with four backups retained.
+
+User-generated codelist content is stored in the core database along with
+information on users and organizations. This is the only source of truth (SOT)
+for the user-generated content, which takes some time to develop if using the
+codelist builder UI. So codelist data loss is costly.
+
+Other databases hold representations of the various coding systems and mappings
+against which a codelist may be developed, displayed, and interpreted. The SOT
+is the source data files provided by a third-party. These are written to once
+then stored read-only. In theory, we can re-run our code against source to
+re-generate the databases, reducing the cost of data loss. In practice, it may
+be difficult to do this in some cases due to the original code or source data
+being unavailable. Some of the source data is stored locally, but not all.
+
+All other material required to deploy the site is under source control.
+
+## Decision
+
+We will back up the core database nightly using scheduled SQLite CLI commands.
+The backup will be stored in the same file system. The backup will be
+compressed. We will retain these backups for 30 days.
+
+We will not back up the coding system databases other than through the
+continuing weekly droplet backups.
+
+## Consequences
+
+The core database can be straightforwardly restored from the local nightly
+backups, limiting core data loss to one day if the droplet file system is not lost.
```

Comment:
It may be difficult to restore the core database from a droplet backup. As sqlite is a single file, I believe that if the droplet backup is taken at the same time as a process is writing to it, then we can't guarantee consistency of the data. It might be worth checking that and in practice, I think it's quite unlikely given the time the droplet backup is likely to be taken. However, that does make the locally stored backups more important (those could be restored from a droplet backup if needed)

Reply:
The sqlite `.backup` command the ADR suggests takes care of this, by taking a full db lock before doing the copy.

Additionally, it handles the case where there are multiple files related to the db, such as when WAL mode is enabled and there is an additional WAL file.

Reply:
[SQLite, in any `journal_mode` ](https://www.sqlite.org/pragma.html#pragma_journal_mode)other than `memory` or `off`, writes transactions to additional files before they are committed. The core OpenCodelists database uses `wal` mode.

[The SQLIte backup API](https://www.sqlite.org/backup.html) ensures that when completed "the backup database contains a consistent and up-to-date snapshot of the original." The nightly backup process described will use this approach. If writes are performed on the source database during the backup process they will slightly delay the process but be included in the backup database file.

If the DigitalOcean snapshot for the backup is taken while a transaction is being written the eventual droplet backup may contain a corrupt or inconsistent database. I don't know if the droplet backup process does or could do something to ensure this cannot happen, such as entering "maintenance mode" before taking the backup. If not, we might have to restore the latest daily backup after restoring the droplet to get the database in a usable state. And it might be prudent to do this in case of usable but inconsistent data.

Note that this would increase the maximum data loss to 7 days plus the lag between taking the last daily backup and taking the weekly snapshot. If nothing else, I suggest we should make that lag short.

Reply:
It seems sensible to use `wal` in the core database as reads may sometimes be much more common than writes and it helps address concurrency concerns. I checked the mode by downloading and restoring a backup locally and querying `pragma journal_mode;` giving `wal`. However I'm a little concerned that I cannot see where this is configured in the repo. I've only looked for a couple of minutes. If it's not captured in code or in manual deployment instructions, it should be.